osu.Server.Queues.ScoreStatisticsProcessor/Stores/BeatmapStore.cs

```diff
@@ -82,29 +82,34 @@ public static async Task<BeatmapStore> CreateAsync(MySqlConnection connection, M
                 return calculator.Calculate(mods);
             }
 
-            BeatmapDifficultyAttribute[]? rawDifficultyAttributes;
-
             LegacyMods legacyModValue = getLegacyModsForAttributeLookup(beatmap, ruleset, mods);
+
             DifficultyAttributeKey key = new DifficultyAttributeKey(beatmap.beatmap_id, (uint)ruleset.RulesetInfo.OnlineID, (uint)legacyModValue);
 
-            if (!attributeCache.TryGetValue(key, out rawDifficultyAttributes))
-            {
-                rawDifficultyAttributes = attributeCache[key] = (await connection.QueryAsync<BeatmapDifficultyAttribute>(
-                    "SELECT * FROM osu_beatmap_difficulty_attribs WHERE `beatmap_id` = @BeatmapId AND `mode` = @RulesetId AND `mods` = @ModValue", new
-                    {
-                        key.BeatmapId,
-                        key.RulesetId,
-                        key.ModValue
-                    }, transaction: transaction)).ToArray();
-            }
+            if (attributeCache.TryGetValue(key, out DifficultyAttributes? difficultyAttributes))
+                return difficultyAttributes;
 
-            if (rawDifficultyAttributes == null || rawDifficultyAttributes.Length == 0)
-                return null;
+            BeatmapDifficultyAttribute[] dbAttribs = (await connection.QueryAsync<BeatmapDifficultyAttribute>(
+                "SELECT * FROM osu_beatmap_difficulty_attribs WHERE `beatmap_id` = @BeatmapId AND `mode` = @RulesetId AND `mods` = @ModValue", new
+                {
+                    key.BeatmapId,
+                    key.RulesetId,
+                    key.ModValue
+                }, transaction: transaction)).ToArray();
 
-            DifficultyAttributes difficultyAttributes = LegacyRulesetHelper.CreateDifficultyAttributes(ruleset.RulesetInfo.OnlineID);
-            difficultyAttributes.FromDatabaseAttributes(rawDifficultyAttributes.ToDictionary(a => (int)a.attrib_id, a => (double)a.value), beatmap);
+            try
+            {
+                if (dbAttribs.Length == 0)
+                    return difficultyAttributes;
 
-            return difficultyAttributes;
+                difficultyAttributes = LegacyRulesetHelper.CreateDifficultyAttributes(ruleset.RulesetInfo.OnlineID);
+                difficultyAttributes.FromDatabaseAttributes(dbAttribs.ToDictionary(a => (int)a.attrib_id, a => (double)a.value), beatmap);
+                return difficultyAttributes;
+            }
+            finally
+            {
+                attributeCache[key] = difficultyAttributes;
+            }
```

Comment:
I believe all of this works but it's intricate enough to be pushing my limits of how much subtlety I'm willing to accept in code, what with all of the juggling of `difficultyAttributes`, implicit null returns of `TryGetValue()`, and `finally` used to coalesce return paths in order to populate `attributeCache`...

Reply:
Hmm, this is the standard way I'd write any caching code (I believe it's done this was elsewhere) but I'm neutral to approaching it differently. Would be be more understandable if the early return did the null-caching inline?

Reply:
The general pattern is standard yes, I'm more talking about the mechanics here.

> Would be be more understandable if the early return did the null-caching inline?

To a degree, but it would also make it mechanically uglier, so I decided I would just put up with this as is.

Reply:
One thing to keep in mind is that this will cache attributes even if they fail, and then they will pass as if nothing went wrong the next time around. There's two places that can fail - either the attributes are not present in the DB which I can't immediately think of a situation in which it should be allowed, or the deserialisation fails which is always a bad sign.